README
Tk::Wizard - Wizard GUI Framework
=================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make all test
   make install

Be careful with the tests: some are currently interactive.

DESCRIPTION

In the context of this namespace, a Wizard is defined as a graphic user
interface (GUI) that presents information, and possibly performs tasks,
step-by-step via a series of different pages.  Pages (or 'screens', or
'Wizard frames') may be chosen logically depending upon user input.

The "Tk::Wizard" module automates a large part of the creation of a
wizard program to collect information and then perform some complex task
based upon it.

The wizard feel is largly based upon the Microsoft(TM,etc) wizard style:
the default is similar to that found in Microsoft Windows 95; a more
Windows 2000-like feel is also supported (see the "-style" entry in
"WIDGET-SPECIFIC OPTIONS". Subclassing the module to provide different
look-and-feel is highly encourage: please see "NOTES ON SUB-CLASSING
Tk::Wizard".

NB: THIS SHOULD BE CONSIDERED ALPHA SOFTWARE.  ALL CONTRIBUTIONS ARE WELCOME!

AUTHOR
	Lee Goddard (lgoddard@cpan.org) based on work Daniel T Hable.


ChangeLog
Tk::Wizard CHANGE LOG
---------------------

Version 1.953, 4 Mar. 2007 (interim release by Martin Thurn)	
	* Installer.pm: use catpath rather than raw slashes

Version 1.952, 27 Feb. 2007 (interim release by Martin Thurn)	

	* 52_Installer.t: does not rely on $ENV{TMP} now
	* Installer.pm: $VERSION now works on linux
	* Wizard.pm: $VERSION now works on linux

Version 1.951, 19 Feb. 2007 (interim release by Martin Thurn)	
	
	* Wizard.pm: All "ButtonAction" callbacks now must really return non-false value in order for operation to continue, as the documentation has always said.  (Previous release seemed to be opposite?)
	* Wizard.pm: new method addFileSelectPage()
	* All tests pass (with some ignorable warnings) on MSWin32 with ActivePerl build 819; not tested on other platforms (yet).
	
Version 1.9451, 19 June 2006

	* Wizard.pm: CloseWindowEventCycle was calling self->destroy but should have called parent->self->destroy if the parent was defined by itself

Version 1.945, 18 June 2006
	
	* Wizard.pm: fixed -resizable
	* Wizard.pm: fixed -width and -height for individual pages
	* Wizard.pm: Withraw the parent during our lifetime
	* Wizard.pm: Assume we are passed a MainWindow like a normal widget; add one if it doesn't exist, to maintain backwards compatability and to be useful in my installer scripts

Version 1.944, 18 May 2006

	* 50_Installer.t - Installer test checks if we're online by using LWP - no ExtUtils to do this...?
	* Wizard.pM: Lee added finishButtonAction again	
	* Wizard.pm: ET#19340 for clothohk, allowed the -height and -width to adjust the window size per frame
	
Version 1.943, 17 May 2006

	* Wizard.pM: Lee added fixed typo in test causing RT #19297 by clothohk
	* Wizard.pM: Lee added destroy to finishButtonAction prototype
	* Wizard.pm: Lee fiddled with CloseWindowEventCycle destroy 
	* Wizard.pm: Lee updated POD to reflect auto-destroy of previous version.
	* Wizard.pm: Lee added -debug.
	* Installer.pm: update to SYNOPSIS from clothohk
	
Version 1.942, 24 April 2006

	* Wizard.pm: added &fix_wait to prevent weird Tk error messages
	* Wizard.pm: call &fix_wait in blank_frame
	* Installer.pm: &fix_wait 
	* Installer.pm: remove progress bars when job is done and replace with -done_text
	* Updated Installer.t to use Test::More

VERSION 1.941, 23 April 2006
	
	* Wizard.pm: Added Martin's patch to by default support Base64 photos in TaskList
	* Wizard.pm: Added to TaskList more photo states added state indicators
	* Wizard.pm: Added POD for 'destroy'
	* Wizard.pm: Remvoed POD for postFinishButtonAction
	* Wizard.pm: Added preFinishButtonAction
	* Wizard.pm: bug fix in NextButtonEventCycle: CloseWindowEventCycle ends logic,
	  rather than falling through to the remainder of the next button actions.
	
VERSION 1.94, 22 April 2006

	* Patch by Martin Thurn: sorry it's late:
	* Wizard.pm: fix a typo "image_idr"
 	* Wizard.pm: Removed image_dir option
 	* Removed binary images in lib/Wizard/images
 	* Added Tk::Wizard::Image to supply Base 64 images
	* Wizard.pm: add a sentence of pod for postNextButtonAction

VERSION 1.92

	* Patch by Martin Thurn: thanks!
	* in Installer.pm, prevent Tk core dump when background is undef
	* in Installer.pm, eliminate the big empty space on the licence_page
	* in Wizard.pm, make all of the user portion (but none of the button/borders) expand when the window is resized

VERSION 1.90, May 2005
	
	Um... had my first child, can't remember what the patch was.
	
VERSION 1.08,

	* Mainly release to get the test suite working again,
	  and totally automated and devoid of net-access requirement
	* added -image_dir to help test suite
	- Martin said images are found in sitelibexp, not installsitearch
	  but that is not the case from the test suite.
	* replaced -wait to test suite for automated testing
	* dir select page for Win32 now tests for fixed drives when -nowarnings ne 1
	* test script for dir selet page -nowarnings=9
	* propogate -wait to all routines based on blank_frame, even dirselect for auto tests
	* weird errors from NextButtonEventCycle when final page has -wait call "forward"
	* simplified Installer test, remvoing dir select
	* Added hashes to Installer warnings
	* Passes all tests on Windoze. Will try Debian on Monday.

VERSION 1.07, patched by Martin Thurn <mthurn@verizon.net>:

	* lib/Tk/Wizard.pm (Show, render_current_page): BUGFIX for DirSelectPage leaving Cwd in an unknown place
	* lib/Tk/Wizard.pm: Fixed [most!] typos in pod
	* lib/Tk/Wizard.pm (page_dirSelect): FIX: if caller dictates height, expand widgets to fit
	* lib/Tk/Wizard.pm (blank_frame): BUGFIX: top image had wrong bgcolor
	* lib/Tk/Wizard.pm (Populate): BUGFIX: -imagepath instead of -topimagepath in 2 places
	* lib/Tk/Wizard.pm (Populate): FIX for some widgets having wrong bgcolor
	* lib/Tk/Wizard.pm (page_dirSelect): clean up extraneous hardcoded references to %ENV
	* lib/Tk/Wizard.pm (page_dirSelect): BUGFIX: missing '-' in front of 'command'
	* lib/Tk/Wizard.pm (Populate): cleaned up the "tag text" (still not perfect, but better)
	* lib/Tk/Wizard.pm (_font_family): more flexible font handling; better default fonts on Unix
	* lib/Tk/Wizard.pm (Populate): images are found in sitelibexp, not installsitearch

VERSION 1.06

	*	Tk::Installer supports downloads;

	*	-continue dropped for semmantically inverse but more apt -wait

	*	Tk::Wizard::Installer::Win32 uses Win32.pm falls back to Win32::OLE

	*	Tk::Wizard - added -wait to blank_page

	*	First and last frames fill frame when -style=>top is specified

	*	DIALOGUE_really_quit referred to "Set Up" which wasn't very generic.
		Now just refers to the wizard.

	*	Bug fixed for default -style.

VERSION 1.038, 11 April 2003

	*	Images install to perlsitearch dir and are
		used by default; added more warnings.

	*	Win32::Installer can now create shortcuts!

VERSION 1.037, 08 April 2003


lib/Tk/Wizard/Installer/Win32.pm
package Tk::Wizard::Installer::Win32;

use vars qw/$VERSION/;

$VERSION = do { my @r = (q$Revision: 1.2 $ =~ /\d+/g); sprintf "%d."."%03d" x $#r, @r };



BEGIN {

	use Carp;

	use Tk::Wizard::Installer;

	use Win32::Shortcut;

	use File::Path;

	require Exporter;

	@ISA = "Tk::Wizard::Installer";

	@EXPORT = ("MainLoop");

	use Cwd;

	require Win32;

	if ($Win32::VERSION lt 0.2){

		eval 'use Win32::OLE'; # autouse is still not very good?

		die "Could not load Win32::OLE: $@" if $@;

	}

	use Win32;

	use Win32::TieRegistry( Delimiter=>"/", ArrayValues=>0 );



}





=head1 NAME



Tk::Wizard::Installer::Win32 - Win32-specific routines for Tk::Wizard::Installer



=head1 DESCRIPTION



All the methods and means of C<Tk::Wizard> and C<Tk::Wizard::Installer>,

plus the below, which are thought to be specific to the Microsoft Windows

platform.



If you are looking for a freeware software installer that is not

dependant upon Perl, try Inno Setup - C<http://www.jrsoftware.org/>. It's

so good, even Microsoft have been caught using it.



If you are looking for a means to update the Windows C<Path> variable,

have a look for I<PathTool.exe>, a tiny Windows 32-bit executable

by Luke Bailey (C<luke@notts.flexeprint.com>). This tool can also be

used to add new, persistant environment variables to the system.



=head1 DEPENDENCIES



	Tk::Wizard

	Tk::Wizard::Installer

	Win32::TieRegistry

	Tk::Hlist

	Win32

	Win32::Shortcut

	File::Path



=head1 METHODS



=head2 METHOD register_with_windows



Registers an application with Windows so that it can be "uninstalled"

using the I<Control Panel>'s I<Add/Remove Programs> dialogue.



An entry is created in the Windows' registry pointing to the

uninstall script path. See C<UninstallString>, below.



Returns C<undef> on failure, C<1> on success.

Does nothing on non-MSWin32 platforms



Aguments are:



=over 4



=item uninstall_key_name



The name of the registery sub-key to be used. This is transparent to the

end-user, but should be unique for all applications.



=item UninstallString



The command-line to execute to uninstall the script.



According to L<Microsoft|http://msdn.microsoft.com/library/default.asp?url=/library/en-us/dnwue/html/ch11d.asp>:



	You must supply complete names for both the DisplayName and UninstallString

	values for your uninstall program to appear in the Add/Remove Programs

	utility. The path you supply to Uninstall-String must be the complete

	command line used to carry out your uninstall program. The command line you

	supply should carry out the uninstall program directly rather than from a

	batch file or subprocess.



The default value is:



	perl -e '$args->{app_path} -u'



This default assumes you have set the argument C<app_path>, and that it

checks and reacts to the the command line switch C<-u>:



	package MyInstaller;

	use strict;

	use Tk::Wizard;

	if ($ARGV[0] =~ /^-*u$/i){

		# ... Have been passed the uninstall switch: uninstall myself now ...

	}

	# ...



Or something like that.



=item QuiteUninstallString



As C<UninstallString> above, but for ... quiet uninstalls.



=item app_path



Please see the entry for C<UninstallString>, above.



=item DisplayName



=item DisplayVersion



=item Size



The strings displayed in the application list of the Add/Remove dialogue.



=item ModifyPath



=item NoRepair NoModify NoRemove



=item EstimatedSize InstallSorce InstallDate InstallLocation



=item AthorizedCDFPrefix Language ProductID



Unknown



=item Comments



=item RegOwner



=item RegCompnay



=item Contact



=item HelpTelephone



=item Publisher



=item URLUpdateInfo



=item URLInfoAbout



=item HelpLink



These are all displayed when the Support Information link

is clicked in the Add/Remove Programs dialogue. The last

should be full URIs.



=back



The routine will also try to add any other paramters to the

registry tree in the current location: YMMV.



=cut



sub register_with_windows { my ($self,$args) = (shift,{@_});

	return 1 if $^O!~/(mswin32|cygwin)/i;

	unless ($args->{DisplayName} and $args->{UninstallString}

		and ($args->{uninstall_key_name} or $args->{app_path})

	){

		die __PACKAGE__."::register_with_windows requires an argument of name/value pairs which must include the keys 'UninstallString', 'uninstall_key_name' and 'DisplayName'";

	}



	if (not $args->{UninstallString} and not $args->{app_path}){

		die __PACKAGE__."::register_with_windows requires either argument 'app_path' or 'UninstallString' be set.";

	}

	if ($args->{app_path}){

		$args->{app_path} = "perl -e '$args->{app_path} -u'";

	}

	my $uninst_key_ref =

	$Registry->{'LMachine/SOFTWARE/Microsoft/Windows/CurrentVersion/Uninstall/'} ->

		CreateKey( $args->{uninstall_key_name} );

	die "Perl Win32::TieRegistry error" if !$uninst_key_ref;



	foreach (keys %$args){

		next if $_ =~ /^(app_path|uninstall_key_name)$/g;

		$uninst_key_ref->{"/$_"} = $args->{$_};

	}

	return 1;

	# return $!? undef : 1;

}





=head1 METHOD page_start_menu



Returns a page (a filled C<Tk::Frame> object) that allows

users to select a location on the Windows "Start Menu",

perhaps to add a shortcut there.



This routine does not currently create the directory in the

I<Start Menu>, nor does it place a link there - see

L</CALLBACK callback_create_shortcut> for that. Rather, the

caller supplies a C<-variable> paramter that is a reference

to a scalar which, once the page is 'run', will contain

either the path to the user's chosen directory, or C<undef>

if the option to not select was chosen.



In addition, when the page is 'run', it places the path to the

current user's I<Start Menu/Programs> directory into the object

field C<startmenu_dir_current>, and the path to the common

I<Start Menu/Programs> in the object field C<startmenu_dir_common>.



The adding of a shortcut is left as an exercise to the reader

until later this year, when I'll get around to finishing this.

Have a look at the C<$mkdir> C<Button> in C<Tk::Wizard::page_dirSelect>.



=over 4



=item -user



Set to C<current>, C<all> or C<both> to list the "Start Menu"

for all users, just the current user, or both. Default is C<both>.

If there exist entries in both the common and current user's

I<Start Menu> with the same name, the entry in the common menu

takes precedence.



=item -variable



A reference to a variable that, when the page is completed,

will contain the directory the user has chosen to create an item in. Note

this is I<not> the full path: see above.



=item -program_group



Name of the directory to create on the start menu, if any.

If defined, this will be appended to any selection the user makes.

Since this is just the GUI part, no directory will actually be

created until C<callback_create_shortcut> is called (though this will

change).



=item -disable_nochoice



Set to prevent the display of the checkbox which

allows the user not to use this feature. See C<-label_nochoice>,

below.



=item -label_nochoice



If the paramter C<-disable_nochoice> has not been set,

C<-label_nochoice> should contains text to use for the label by the

checkbox which disables choices on

this page and causes the page to set the C<-variable> paramter to

C<undef>. Default text is C<Do not create a shortcut on the Start Menu>



=item -listHeight



Height of the list box, default is C<10> but you may

vary this if your C<-text> attribute takes up more or less much room.



=back



Accepts the Standard Options that are common to the C<HList> and

C<Label> widgets, but does not accept aliases:



	-relief -highlightthickness -background -borderwidth -cursor

	-highlightcolor -foreground -font



You can supply the common Wizard page options:



	-title -subtitle -text



This method will initially attempt to use F<Win32.pm>; failing

that, it will attempt to use a Windows Scripting Host object

created via C<Win32::OLE>. If both fail (WSH only existing by

default in Win98 and above), the routine will return C<undef>,

rather than a page frame object. This may not be ideal but works

for me - suggestions welcomed for a better idea.



=cut



#sub addStartMenuPage { my ($self,$args) = (shift,{@_});

#	return $self->addPage( sub { $self->page_start_menu($args)  } );

#}





sub page_start_menu { my ($self) = (shift);

	if (ref $_[0] eq 'HASH'){

		$args = shift

	} else {

		$args = {@_};

	}

	local *DIR;

	my (@list,$dir);

	my $cwd = cwd;

	my $do_set = 1;

	croak "You must set -variable parameter"

		unless exists $args->{-variable};

	$args->{-background} = 'white' unless exists $args->{-background};

	$args->{-relief} = 'sunken' unless exists $args->{-relief};

	$args->{-border} = 1 unless exists $args->{-border};

	$args->{-listHeight} = 10 unless exists $args->{-listHeight};

	$args->{-title} = "Create Shortcuts" unless exists  $args->{-title};

	$args->{-subtitle} = "Please select where to place an icon on the start menu" unless exists  $args->{-subtitle};

	$args->{-label_nochoice} = "Do not create a shortcut on the Start Menu" unless exists $args->{-label_nochoice};

	$self->{-program_group} = $args->{-program_group};



	my $common_args; # formatting

	$common_args->{-background} = $args->{-background} if exists $args->{-background};

	$common_args->{-relief} = $args->{-relief} if exists $args->{-relief};

	$common_args->{-highlightthickness} = $args->{-highlightthickness} if exists $args->{-highlightthickness};

	$common_args->{-borderwidth} = $args->{-borderwidth} if exists $args->{-borderwidth};

	$common_args->{-cursor} = $args->{-cursor} if exists $args->{-cursor};

	$common_args->{-highlightcolor} = $args->{-highlightcolor} if exists $args->{-highlightcolor};

	$common_args->{-foreground} = $args->{-foreground} if exists $args->{-foreground};

	$common_args->{-font} = $args->{-font} if exists $args->{-font};

	# Don't pass these to other modules

	my $variable = $args->{-variable};

	delete $args->{-variable};

	my $group = $args->{-program_group};

	delete $args->{-program_group};



	my $frame = $self->blank_frame(%$args);



	if ($Win32::VERSION gt 0.1999999){

		$self->{startmenu_dir_current} =

			eval('Win32::GetFolderPath(Win32::CSIDL_STARTMENU)')

			. '\Programs';

		$self->{startmenu_dir_common}  =

			eval('Win32::GetFolderPath(Win32::CSIDL_COMMON_STARTMENU)')

			. '\Programs';

	}

	# The above may not work if non-standard/non-English setup, so:

	if (not $self->{startmenu_dir_current}

	or not $self->{startmenu_dir_common}){

		my $WshShell = eval 'Win32::OLE->CreateObject("WScript.Shell")';

		if (ref $WshShell eq 'Win32::OLE'){

			$self->{startmenu_dir_current} = $WshShell->SpecialFolders(17);

			$self->{startmenu_dir_common}  = $WshShell->SpecialFolders(2);

		} else {

			warn "Could not find special folders using Win32 or OLE!";

			return undef;

		}

	}

	if ($args->{-user} eq 'current'){

		$dir = [$self->{startmenu_dir_current}];

	} elsif ($args->{-user} eq 'all') {

		$dir = [$self->{startmenu_dir_common}];

	} else {

		$dir = [

			$self->{startmenu_dir_common},

			$self->{startmenu_dir_current}

		];

	}

	$$variable = @$dir[0]."\\".$group;

	foreach my $dodir (@$dir){

		chdir $dodir;

		opendir DIR,$dodir or croak "I couldn't open the start menu ($dodir)";

		push @list, grep {-d && !/^\.\.?$/ && s/^(.*)$/$dodir\\$1/} sort readdir DIR;

		close DIR;

	}

	chdir $cwd;

	my $listbox = $frame->Scrolled("HList",

		-scrollbars => "osoe",

		-selectmode => 'single',

		-height		=> $args->{-listHeight},

		-itemtype   => 'text',

		-separator  => '',

		-browsecmd	=> sub {

			$$variable = shift;

			$$variable .= "\\".$group if $group;

		},

		%$common_args

	)->pack(qw/ -expand 1 -fill x -padx 10 -pady 10 /);



	foreach my $i (@list){

		#my ($t) = $i =~ m/([^\\\/]+)$/;

		my $t = $i;

		unless ($listbox->info("exists", $t)){

			$listbox->add( $t,

				-text => $t,

				-data => $i,

			);

		}

	}



	my $text	= $frame->Label( # Entry(

		-justify		=> 'left',

		-textvariable	=> $variable,

		-anchor => 'w',

		%$common_args

	)->pack(

		-side=>'top',-anchor=>'w',-expand=>1,-fill=>"x", -padx=>10,

	);

	my $button;



	unless ($args->{-disable_nochoice}){

		$button = $frame->Checkbutton(

			-text	=> $args->{-label_nochoice},

			-anchor	=> 'w',

			-command => sub {

				$do_set = !$do_set;

				if (not $do_set){	# De-activate

					$$variable = '';

					$text->configure(-background=>$self->cget(-background));

					$listbox->configure(-background=>$self->cget(-background));

				} else {

					$$variable = $listbox->info('anchor');

					$$variable .= "\\".$group if $group;

					$text->configure(-background=>$args->{-background});

					$listbox->configure(-background=>$args->{-background} );

				}

			},

		)->pack(qw/-side left -anchor w -padx 2 -pady 2/);

	};

	return $frame;

}





=head1 CALLBACK callback_create_shortcut



A convenience interface to C<Win32::Shortcut> method that creates a shortcut

at the path specified. Parameters are pretty much what you see when

you right-click a shortcut:



=over 4



=item -save_path



The location at which the shortcut should be saved.

This should be the full path including filename ending

in C<.lnk>.



The filename minus extension will be visible in the shortcut.

If the C<-program_group> parameter was passed to

C<METHOD page_start_menu>, the directory it refers to

will be included in the save path you supply.To avoid

this, either C<undef>ine the object field C<-program_group>,

or supply the paramter C<-no_program_group>.



=item -no_program_group>



See C<-save_path>, above.



=item -target



The shortcut points to this file, directory, or URI -

see notes for C<-save_path>, above.



=item -workingdir



The working directory for the C<-target>, above.



=item -description



This is what you see when you mouse-over a shortcut

in more "modern" (Win2k/ME+) Windows.



=item -iconpath



Path to the icon file - an C<.exe>, C<.dll>, C<.ico> or

other acceptable format.



=item -iconindex



Index of the icon in the file if a C<.exe> or C<.dll>.



=item -arguments



Um... it's the second parameter in Win32::Shortcut::Set -

could well be parameters for the target, but I'm too much of

a rush to check. XXX



=item -show



Whether the C<-target>, above, should be

started maximized or minimized. Acceptable values are

the constants:



    SW_SHOWMAXIMIZED SW_SHOWMINNOACTIVE SW_SHOWNORMAL



=item -hotkey



Key combination to activate the shortcut. Probably looks

something like C<ctrl+t>.



=back



On success, returns the C<-save_path>; on failure, C<undef>.



=cut



sub callback_create_shortcut { my ($self,$args) = (shift,{@_});

	croak "-target is required (you gave ".(join", ",keys %$args).")"	 unless defined $args->{-target};

	croak "-save_path is required  (you gave ".(join", ",keys %$args).")"		unless defined $args->{-save_path};

	$args->{-arguments} 	= '' unless exists $args->{-arguments};

	$args->{-description}	= '' unless exists $args->{-description};

	$args->{-show}			= '' unless exists $args->{-show};

	$args->{-hotkey} 		= '' unless exists $args->{-hotkey};

	$args->{-iconpath} 		= '' unless exists $args->{-iconpath};

	$args->{-iconindex}		= '' unless exists $args->{-iconindex};

	undef $self->{-program_group} if exists $args->{-no_program_group};

	if (exists $self->{-program_group}

	and exists $args->{-save_path}

	){

		my ($base,$file) = $args->{-save_path} =~ /(.*?)([^\/\\]+)$/;

		$base .= "\\".$self->{-program_groups};

		mkpath $base if !-e $base;

		$args->{-save_path} = $base."\\".$file;

		$args->{-save_path} =~ s/[\\\/]+/\\/g;

	}

	if ($args->{-target} =~ /^(ht|f)tp:\/\//

		and $args->{-save_path} !~ /\.uri$/i){

		croak "Internet shortcuts require a .uri ending!"

	}

	if ($args->{-target} !~ /^(ht|f)tp:\/\//

		and $args->{-save_path} =~ /\.uri$/i){

		croak "Only internet shortcuts require a .uri ending!"

	}

	my $s = new Win32::Shortcut;

	$s->Set(

		$args->{-target},

		$args->{-arguments},

		$args->{-workingdir},

		$args->{-description},

		$args->{-show},

		$args->{-hotkey},

		$args->{-iconpath},

		$args->{-iconindex},

	);

	my $r = $s->Save($args->{-save_path})?

		$args->{-save_path} : undef;

	$s->Close;

	return $r;

}



=head1 CALLBACK callback_create_shortcut



Convenience method to create multiple shortcuts at once.

Supply an array of hashes, each hash being arguments

to supply to C<callback_create_shortcut>.



Returns an array or reference to an array that contains

the reults of the shortcut creation.



See L<CALLBACK callback_create_shortcut>.



=cut



sub callback_create_shortcuts { my $self = shift;

	my @paths;

	foreach (@_){

		confess "Not a hash reference" unless ref $_ eq 'HASH';

		$self->callback_create_shortcut(%$_);

	}

	return wantarray? @paths : \@paths;

}













=head1 DIALOUGE METHOD DIALOGUE_really_quit



Refers to the Instllaer, rather than the Wizard.



=cut



sub DIALOGUE_really_quit { my $self = shift;

	return 0 if $self->{nextButton}->cget(-text) eq $LABELS{FINISH};

	unless ($self->{really_quit}){

		my $button = $self->parent->messageBox('-icon' => 'question', -type => 'yesno',

		-default => 'no', -title => 'Quit The Wizard?',

		-message => "The Installer has not finished running.\n\nIf you quit now, the installation will be incomplete.\n\nDo you really wish to quit?");

		$self->{really_quit} = lc $button eq 'yes'? 1:0;

	}

	return !$self->{really_quit};

}





 1;

__END__



=head1 CAVEATS AND BUGS



* Error going backwards into a C<addStartMenuPage>.



=head1 CHANGES



Please see the file F<CHANGES.txt> included with the distribution.



=head1 AUTHOR



Lee Goddard (lgoddard@cpan.org).



=head1 SEE ALSO



Tk::Wizard; Tk::Wizard::Installer;

Win32::GetFolderPath();

Win32::Shortcut;

Win32::OLE.



=head1 KEYWORDS



Wizard; set-up; setup; installer; uninstaller; install; uninstall; Tk; GUI; windows; win32; registry; shortcut;



=head1 COPYRIGHT



Copyright (C) Lee Goddard, 11/2002 ff.



Distributed under the same terms as Perl itself.



Makefile.PL
use ExtUtils::MakeMaker;

my $prereq = {
	"Tk" => 0.1,
	"Tk::ProgressBar"	=> 0.1,
	"Tk::LabFrame" => 0.1,
	"File::Path" => 0.1,
	"File::Path"	=> 0,
	"Tk::DirTree" => 0,
	"Tk::ROText"  => 0,
};

if ($^O =~ m/win/i) {
	$prereq->{'Win32::TieRegistry'} = 0.1;
	$prereq->{'Win32::Shortcut'} = 0.03;
	eval 'require Win32;';
	if (! $@) {
		eval '$_ = $Win32::VERSION;';
		if (! $@ && ($_ < 0.2)) {
			eval 'require Win32::OLE;';
			warn "Win32.pm version is $_ (less than 0.2), therefore we will use Win32::OLE.  (This requires Windows Scripting Host which is present on Win98 and higher.)\n";
			$prereq->{'Win32::OLE'} = 0.1;
		} else {
			warn "Using Win32 for special folders...\n";
			$prereq->{'Win32'} = 0.2;
		}
	} else {
		die "There appears to be something wrong with your Perl installation. You are on a Windows platform without the core Win32 module. You can still use this module, but parts you will not be able to create shortcuts to the windows special folders.";
	}
}

WriteMakefile(
	'ABSTRACT' => 'GUI for step-by-step interactive logical process',
	'DISTNAME' => 'Tk-Wizard-Bases',
	'NAME' => 'Tk::Wizard',
	'VERSION'	=> 1.953,
	'PREREQ_PM' => $prereq,
	(
		$] >= 5.005 ?    ## Add these new keywords supported since 5.005
		(#ABSTRACT_FROM => 'Wizard.pm', # retrieve abstract from module
		AUTHOR => 'cpan -at- lgoddard -dot- net>') : ()
	),
);

__END__
